build/nested-cluster-base/Dockerfile

```diff
@@ -0,0 +1,76 @@
+# Copyright 2017 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM gcr.io/google-containers/debian-base-amd64:0.3
+MAINTAINER qlee@google.com
+
+ENV container docker
+
+ARG DEBIAN_FRONTEND=noninteractive
+RUN clean-install apt-utils \
+    apt-transport-https \
+    ca-certificates \
+    curl \
+    gnupg2 \
+    software-properties-common \
+    curl \
+    lxc \
+    iptables \
+    iproute2 \
+    socat \
+    util-linux \
+    mount ebtables \
+    ethtool \
+    vim \
+    bash \
+    procps \
+    less \
+    kmod \
+    systemd \
+    systemd-sysv \
+    jq
+
+# Install docker.
+RUN curl -fsSL https://download.docker.com/linux/debian/gpg | apt-key add - && apt-key fingerprint 0EBFCD88 && add-apt-repository \
+    "deb [arch=amd64] https://download.docker.com/linux/debian \
+    jessie \
+    stable"
+RUN clean-install docker-ce
+
+# Install the magic wrapper.
+COPY dind /bin/dind
+COPY dockerd-entrypoint.sh /bin/dockerd-entrypoint.sh
+
+# Remove unwanted systemd services.
+RUN for i in /lib/systemd/system/sysinit.target.wants/*; do [ "${i##*/}" = "systemd-tmpfiles-setup.service" ] || rm -f "$i"; done; \
+  rm -f /lib/systemd/system/multi-user.target.wants/*;\
+  rm -f /etc/systemd/system/*.wants/*;\
+  rm -f /lib/systemd/system/local-fs.target.wants/*; \
+  rm -f /lib/systemd/system/sockets.target.wants/*udev*; \
+  rm -f /lib/systemd/system/sockets.target.wants/*initctl*; \
+  rm -f /lib/systemd/system/basic.target.wants/*;\
+  rm -f /lib/systemd/system/anaconda.target.wants/*;
+
+# We don't want systemd to start docker for us.
+RUN find / -name docker.conf -exec rm {} \;
+RUN find / -name docker.service -exec rm {} \;
+RUN find / -name docker.socket -exec rm {} \;
+
+# TODO(Q-Lee): SIGRTMIN+3 is the shutdown signal for systemd. Investigate why "STOPSIGNAL SIGRTMIN+3" breaks.
+
+# Define additional metadata for our image.
+VOLUME /var/lib/docker
+VOLUME /var/lib/kubelet
+
```

Comment:
There are many layers being created here. Have you consider using multi-stage builds?
What is the use of this image exactly? Is it used as a base image for another one or it's a build image of some sort?

Reply:
There's a single base image built with docker and pre-published periodically that does all of the non-hermetic and non-k8s-source-specific bits. The rest of the layers are built with bazel from the current source on demand. This should work well for cheaper integration style testing. 

Eventually we can fold the build into the existing bazel-build presubmit and have various tests run using the artifacts. Though many tests will still be better off on a more realistic cluster like the gce-e2e / kops-e2e jobs or the node tests.

The base image pattern comes from here: https://github.com/bazelbuild/rules_docker

Reply:
I am just worried about yet another complicated build+test scenario, just like all the ones we already have, it'd be great if we have take care of making it simple and easy to debug, once we are starting with a (mostly) clean sheet.
Right now this particular dockerfile introduces many layers and I am not sure whether that's a build cache optimisation or no particular attention had been paid to neither build cache nor final image size.